writer.go

```diff
@@ -0,0 +1,145 @@
+// Package sstable implements the reading and write of SSTable files.
+//
+// SSTable file format is defined here:
+// https://www.igvita.com/2012/02/06/sstable-and-log-structured-storage-leveldb/
+// In particular, a basic SSTable file contains a sequence of records,
+// where each record consists of:
+//
+//  1. key-size in 4 bytes of little-endian uint32
+//  2. key
+//  3. value-size in 4 bytes of little-endian uint32
+//  4. value
+//
+// An SSTable file can optionally has the index data structured
+// appended to records:
+//
+//  1. a separate of 4 bytes of little-endian encoding of 0xffffffff
+//  2. number of unique keys as 4 bytes of little-endian encoding
+//  3. a sequence of key and offsets of values of that key:
+//     1. key-size in 4 bytes of little-endian uint32
+//     2. key
+//     3. number of records of that key in 4 bytes of little-endian uint32
+//     4. offsets in the SSTable file of all these records
+//  4. index-end flag as
+//     1. 4 bytes of little-endiean encoding of 0xffffffff
+//     2. 8 bytes of the offset of the index.
+//
+// Note that the following are all valid SSTable files:
+//
+//  1. records
+//  2. records + separator 0xffffffff
+//  3. records + separator 0xffffffff + index + index-end 0xffffffff
+//
+package sstable
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+const (
+	separator        uint32 = 0xffffffff
+	indexEndFlagSize int64  = 12 // 0xffffff and the 8-byte offset
+)
+
+// Writer constructs an SSTable file.  It is not thread-safe.
+type Writer struct {
+	io.Writer
+	index  map[string][]int64
+	offset int64
+}
+
+func Create(w io.Writer) *Writer {
+	return &Writer{
+		Writer: w,
+		index:  make(map[string][]int64),
+		offset: 0}
+}
+
+func (ss *Writer) Put(key, value string) error {
```

Comment:
Maybe not related to this PR, but when using Put, we need to make sure that the keys are sorted when putting into the SSTable.